File: buy_suit/python/buy-socket-http1.py
# -*- coding:utf-8 -*-

# https://github.com/lllk140

from urllib.parse import urlencode
from urllib.parse import urlsplit
from urllib.parse import unquote
import hashlib
import socket
import uuid
import time
import ssl


class Tools(object):
    def __init__(self):
        super(Tools, self).__init__()

    @staticmethod
    def BiliTraceId(_time: int | float | str = None):
        _time = float(_time) if _time else time.time()
        back6 = hex(round(_time / 256))
        front = str(uuid.uuid4()).replace("-", "")
        _data1 = front[6:] + back6[2:]
        _data2 = front[22:] + back6[2:]
        return f"{_data1}:{_data2}:0:0"

    @staticmethod
    def AddFormDataSign(form_data: str):
        app_sec = "560c52ccd288fed045859ed18bffd973"
        form_data_sec = f"{form_data}{app_sec}"
        md5_hashlib = hashlib.md5()
        md5_hashlib.update(form_data_sec.encode())
        sign = md5_hashlib.hexdigest()
        return form_data + f"&sign={sign}"

    @staticmethod
    def BuildFormData(sale_time, **kwargs):
        __add_month = kwargs.get("add_month", "-1")
        __buy_num = kwargs.get("buy_num", "1")
        __coupon_token = kwargs.get("coupon_token", "")
        __f_source = kwargs.get("f_source", "shop")
        __shop_from = kwargs.get("shop_from", "feed.card")

        __statistics = kwargs["statistics"]
        __access_key = kwargs["access_key"]
        __item_id = kwargs["item_id"]
        __app_key = kwargs["appkey"]
        __csrf = kwargs["csrf"]

        form_data_text = urlencode({
            "access_key": __access_key,
            "add_month": str(__add_month),
            "appkey": str(__app_key),
            "buy_num": str(__buy_num),
            "coupon_token": str(__coupon_token),
            "csrf": str(__csrf),
            "currency": "bp",
            "disable_rcmd": "0",
            "f_source": str(__f_source),
            "from": str(__shop_from),
            "from_id": "",
            "item_id": str(__item_id),
            "m_source": "",
            "platform": "android",
            "statistics": __statistics,
            "ts": str(sale_time)
        })
        return form_data_text

    @staticmethod
    def ParseHttpMessage(content: bytes) -> tuple[dict, dict, dict]:
        message_content: list = content.split(b"\r\n")

        request: list = message_content[0].split(b" ")
        url_query: bytes = urlsplit(request[1]).query
        p = [i.split(b"=") for i in url_query.split(b"&")]
        p2 = [[ii.decode() for ii in i] for i in p]
        params = {unquote(i[0]): unquote(i[1]) for i in p2}

        headers_content = message_content[1:len(message_content) - 2]
        h = [i.split(b": ") for i in headers_content]
        h2 = [i if len(i) == 2 else [i[0], b""] for i in h]
        h3 = [[ii.decode() for ii in i] for i in h2]
        headers = {unquote(i[0]): i[1] for i in h3}

        __key = "Cookie" if "Cookie" in headers else "cookie"
        cookies_content: str = headers.get(__key)
        c1: list = cookies_content.split("; ")
        c2 = [i.split("=") for i in c1]
        cookies = {i[0]: i[1] for i in c2}

        return params, headers, cookies


class SuitValue(Tools):
    def __init__(self, http_message: bytes, sale_time: int, **kwargs):
        super(Tools, self).__init__()

        self.host = kwargs.get("host", "api.bilibili.com")

        params, headers, cookies = self.ParseHttpMessage(http_message)

        kwargs.update({"csrf": params["csrf"]})
        kwargs.update({"appkey": params["appkey"]})
        kwargs.update({"item_id": params["item_id"]})
        kwargs.update({"access_key": params["access_key"]})
        kwargs.update({"statistics": params["statistics"]})

        # 表单
        form_data_content = self.BuildFormData(sale_time, **kwargs)
        form_data: str = self.AddFormDataSign(form_data_content)

        # 完整报文
        __message = self.BuildMessage(sale_time, headers, form_data)
        self.message_header = __message[:-1]
        self.message_body = __message[-1:]

    def BuildMessage(self, sale_time: int, headers: dict, form_data: str):
        message = "POST /x/garb/v2/trade/create HTTP/1.1\r\n"
        message += f"native_api_from: {headers['native_api_from']}\r\n"
        message += f"Cookie: {headers['Cookie']}\r\n"
        message += f"Accept: {headers['Accept']}\r\n"
        message += f"Referer: {headers['Referer']}\r\n"
        message += f"env: {headers['env']}\r\n"
        message += f"APP-KEY: {headers['APP-KEY']}\r\n"
        message += f"User-Agent: {headers['User-Agent']}\r\n"
        message += f"x-bili-trace-id: {self.BiliTraceId(sale_time)}\r\n"
        message += f"x-bili-aurora-eid: {headers['x-bili-aurora-eid']}\r\n"
        message += f"x-bili-mid: {headers['x-bili-mid']}\r\n"
        message += f"x-bili-aurora-zone: {headers['x-bili-aurora-zone:']}\r\n"
        message += f"Content-Type: application/x-www-form-urlencoded; charset=utf-8\r\n"
        message += f"Content-Length: {str(len(form_data))}\r\n"
        message += f"Host: {self.host}\r\n"
        message += f"Connection: {headers['Connection']}\r\n"
        message += f"Accept-Encoding: {headers['Accept-Encoding']}\r\n\r\n"
        return str(message + form_data).encode()


class SuitBuy(SuitValue):
    def __init__(self, http_message: bytes, sale_time: int, **kwargs):
        super(SuitBuy, self).__init__(http_message, sale_time, **kwargs)

    def CreateTlsConnection(self, port: int = 443, **kwargs) -> ssl.SSLSocket:
        context = ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
        context.purpose = ssl.Purpose.SERVER_AUTH
        context.verify_mode = ssl.CERT_REQUIRED
        context.check_hostname = True
        context.load_default_certs()
        _connection = socket.create_connection((self.host, port))
        kwargs.update({"server_hostname": self.host})
        connection = context.wrap_socket(_connection, **kwargs)
        return connection

    def SendMessageHeader(self, client: ssl.SSLSocket):
        return client.send(self.message_header)

    def SendMessageBody(self, client: ssl.SSLSocket):
        return client.send(self.message_body)

    @staticmethod
    def ReceiveResponse(client: ssl.SSLSocket, length=4096) -> bytes:
        return client.recv(length)

    def demo(self, port=443, **kwargs):
        # 创建连接
        client = self.CreateTlsConnection(port, **kwargs)

        s = time.time()

        # 发送报文
        self.SendMessageHeader(client)
        self.SendMessageBody(client)

        # 接收响应
        response = self.ReceiveResponse(client)

        e = time.time()

        print(response.decode())
        print(e - s)
        client.close()
        return response, e - s


def main():
    sale_time = 1665889008

    suit_buy = SuitBuy(
        http_message=open(r"../http-message/HTTP1.1Message.txt", "rb").read(),
        sale_time=sale_time,

        # 可选
        add_month="-1",
        buy_num="1",
        coupon_token="",
        host="api.bilibili.com",
        f_source="shop",
        shop_from="feed.card",
    )

    # 演示
    response, run_time = suit_buy.demo()
    print(response, run_time)

    # 跳出本地计时器后
    # client = suit_buy.CreateTlsConnection()
    # suit_buy.SendMessageHeader(client)

    # 等待服务器计时退出
    # suit_buy.SendMessageBody(client)
    # response = suit_buy.ReceiveResponse(client)

    # print(response.decode())
    
    # 关闭连接
    # client.close()


if __name__ == '__main__':
    main()

File: buy_suit/golang/buy-socket-http1.go
package main

import (
	"bytes"
	"crypto/md5"
	"crypto/tls"
	"encoding/hex"
	"fmt"
	"math/rand"
	"net/url"
	"os"
	"sort"
	"strconv"
	"strings"
	"time"
)

type Config struct {
	saleTime    int64
	addMonth    int64
	buyNum      int64
	host        string
	couponToken string
	fSource     string
	shopFrom    string
}

// buildRandomText
// 生成一个随机字符串, 代替原uuid4
func buildRandomText(size int64) string {
	var KeyString = "abcdef0123456789"
	var RandomBytes bytes.Buffer
	rand.Seed(time.Now().UnixNano())
	for i := 0; i < int(size); i++ {
		RandInt := rand.Intn(len(KeyString))
		RandomBytes.WriteByte(KeyString[RandInt])
	}
	return RandomBytes.String()
}

// urlEncode
// 表单转url编码
func urlEncode(FormData map[string]string) string {
	var ParamList []string
	for key, value := range FormData {
		var Key = url.QueryEscape(key)
		var Value = url.QueryEscape(value)
		var param = fmt.Sprintf("%v=%v", Key, Value)
		ParamList = append(ParamList, param)
	}
	sort.Strings(ParamList)
	return strings.Join(ParamList, "&")
}

// ------------------------------------------------------------------------

// BiliTraceId
// 生成TraceId
func BiliTraceId(Time int64) string {
	var back6 = strconv.FormatInt(Time, 16)
	var RandomText = buildRandomText(32)
	var a, b = RandomText[6:] + back6, RandomText[22:] + back6
	return fmt.Sprintf("%v:%v:0:0", a, b)
}

// AddFormDataSign
// 为表单添加sign
func AddFormDataSign(FormData string) string {
	var AppSec = "560c52ccd288fed045859ed18bffd973"
	var FormDataSec = fmt.Sprintf("%v%v", FormData, AppSec)
	var Md5Hash = md5.New()
	Md5Hash.Write([]byte(FormDataSec))
	var SignUint8 = Md5Hash.Sum(nil)
	var FormDataSign = hex.EncodeToString(SignUint8)
	return fmt.Sprintf("%v&sign=%v", FormData, FormDataSign)
}

// parseHttpMessage
// 解析http1.1报文数据
func parseHttpMessage(content []byte) (map[string]string, map[string]string) {
	var MessageList = strings.Split(string(content), "\r\n")

	var MessageMap = make(map[string]string)
	for _, value := range MessageList[1 : len(MessageList)-2] {
		var ValueList = strings.Split(value, ": ")
		if len(ValueList) != 2 {
			ValueList = []string{ValueList[0], ""}
		}
		var key = strings.ToLower(ValueList[0])
		MessageMap[key] = ValueList[1]
	}

	var ValueMap = make(map[string]string)
	var Url = strings.Split(MessageList[0], " ")[1]
	var Params = strings.Split(Url, "?")[1]
	var ParamsList = strings.Split(Params, "&")
	for _, value := range ParamsList {
		var ValueList = strings.Split(value, "=")
		if len(ValueList) != 2 {
			ValueList = []string{ValueList[0], ""}
		}
		var key, _ = url.QueryUnescape(ValueList[0])
		var _value, _ = url.QueryUnescape(ValueList[1])
		ValueMap[key] = _value
	}

	return MessageMap, ValueMap
}

// BuildFormData
// 生成表单
func BuildFormData(params map[string]string, config *Config) string {
	var FormData = map[string]string{
		"access_key":   params["access_key"],
		"add_month":    strconv.FormatInt((*config).addMonth, 10),
		"appkey":       params["appkey"],
		"buy_num":      strconv.FormatInt((*config).buyNum, 10),
		"coupon_token": (*config).couponToken,
		"csrf":         params["csrf"],
		"currency":     "bp",
		"disable_rcmd": "0",
		"f_source":     (*config).fSource,
		"from":         (*config).shopFrom,
		"from_id":      "",
		"item_id":      params["item_id"],
		"m_source":     "",
		"platform":     "android",
		"statistics":   params["statistics"],
		"ts":           strconv.FormatInt((*config).saleTime, 10),
	}
	return AddFormDataSign(urlEncode(FormData))
}

// BuildMessage
// 生成报文
func BuildMessage(config *Config, headers map[string]string, formData string) []byte {
	var message = "POST /x/garb/v2/trade/create HTTP/1.1\r\n"
	message += fmt.Sprintf("native_api_from: %v\r\n", headers["native_api_from"])
	message += fmt.Sprintf("Cookie: %v\r\n", headers["cookie"])
	message += fmt.Sprintf("Accept: %v\r\n", headers["accept"])
	message += fmt.Sprintf("Referer: %v\r\n", headers["referer"])
	message += fmt.Sprintf("env: %v\r\n", headers["env"])
	message += fmt.Sprintf("APP-KEY: %v\r\n", headers["app-key"])
	message += fmt.Sprintf("User-Agent: %v\r\n", headers["user-agent"])
	message += fmt.Sprintf("x-bili-trace-id: %v\r\n", BiliTraceId((*config).saleTime))
	message += fmt.Sprintf("x-bili-aurora-eid: %v\r\n", headers["x-bili-aurora-eid"])
	message += fmt.Sprintf("x-bili-mi: %v\r\n", headers["x-bili-mid"])
	message += fmt.Sprintf("x-bili-aurora-zone: %v\r\n", headers["x-bili-aurora-zone"])
	message += "Content-Type: application/x-www-form-urlencoded; charset=utf-8\r\n"
	message += fmt.Sprintf("Content-Length: %v\r\n", len(formData))
	message += fmt.Sprintf("Host: %v\r\n", config.host)
	message += fmt.Sprintf("Connection: %v\r\n", headers["connection"])
	message += fmt.Sprintf("Accept-Encoding: %v\r\n\r\n", headers["accept-encoding"])
	return []byte(message + formData)
}

// BuildAll
// 生成所有
func BuildAll(filePath string, config *Config) ([]byte, []byte) {
	var Message, _ = os.ReadFile(filePath)
	var MessageMap, ValueMap = parseHttpMessage(Message)
	var FormData = BuildFormData(ValueMap, config)
	var HttpMessage = BuildMessage(config, MessageMap, FormData)
	return HttpMessage[:len(HttpMessage)-1], HttpMessage[len(HttpMessage)-1:]
}

// CreateTlsConnection
// 创建连接
func CreateTlsConnection(config *Config) *tls.Conn {
	var TlsConfig *tls.Config = &tls.Config{
		InsecureSkipVerify: false,
		ServerName:         config.host,
		MinVersion:         tls.VersionTLS12,
		MaxVersion:         tls.VersionTLS12,
		ClientAuth:         tls.RequireAndVerifyClientCert,
	}
	var adder = fmt.Sprintf("%v:443", config.host)
	var client, err = tls.Dial("tcp", adder, TlsConfig)
	if err != nil {
		fmt.Printf("%v\n", err)
		os.Exit(1)
	}
	return client
}

// SendMessage
// 发送请求
func SendMessage(client *tls.Conn, body []byte) {
	_, _ = client.Write(body)
}

// ReceiveResponse
// 接收响应
func ReceiveResponse(client *tls.Conn) []byte {
	var result = make([]byte, 8192)
	var length, _ = client.Read(result)
	return result[:length]
}

func main() {

	// Config都是必要
	// 一般只要改saleTim就行
	var saleTime = 1666427455
	var filePath = "./buy_suit/http-message/HTTP1.1Message.txt"
	var config = new(Config)
	(*config).saleTime = int64(saleTime)
	(*config).host = "api.bilibili.com"
	(*config).shopFrom = "feed.card"
	(*config).fSource = "shop"
	(*config).buyNum = 1
	(*config).addMonth = -1
	(*config).couponToken = ""

	var header, body = BuildAll(filePath, config)
	//fmt.Printf("%v\n", string(header))
	//fmt.Printf("%v\n", string(body))

	// 跳出本地计时器
	var client = CreateTlsConnection(config)

	var s = time.Now().UnixNano() / 1e6

	SendMessage(client, header) // 发送n-1的内容

	// 跳出服务器计时器
	SendMessage(client, body)              // 发送剩余的内容
	var response = ReceiveResponse(client) // 接收响应

	var e = time.Now().UnixNano() / 1e6

	fmt.Printf("%v\n", string(response))
	fmt.Printf("耗时:%vs\n", e-s)
}
